Adding isin function for multidimensional arrays

brsr opened the pull request:
This resolves #8331. I had initially started the request there, and there was [discussion on the mailing list](https://mail.scipy.org/pipermail/numpy-discussion/2016-December/076305.html), in which everyone seemed pretty happy with the idea. I hadn't realized that there was a separate in1d method for masked arrays, so this includes an isin function for masked arrays too.

brsr committed:
Adding isin function for multidimensional arrays

endolith commented:
I was going to submit this, too, but just using the existing function:

```
def in_nd(ar1, ar2, assume_unique=False, invert=False):
    """
        ...
    """
    out = np.in1d(ar1, ar2, assume_unique=False, invert=False)
    return out.reshape(ar1.shape)
```

endolith commented on numpy/lib/arraysetops.py:
```diff
@@ -380,16 +381,16 @@ def setxor1d(ar1, ar2, assume_unique=False):
+    Test whether each element of an array is present in a second array.
 
-    Returns a boolean array the same length as `ar1` that is True
+    Returns a boolean array the same shape as `ar1` that is True
```
"Returns a boolean array **of** the same shape"

endolith commented on numpy/lib/arraysetops.py:
```diff
@@ -399,70 +400,74 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
-    >>> test = np.array([0, 1, 2, 5, 0])
-    >>> states = [0, 2]
-    >>> mask = np.in1d(test, states)
+    >>> test = np.array([0, 2, 4, 6]).reshape([2, 2])
```
Why not 

    test = np.array([[0, 2], [4, 6]])

?

eric-wieser commented:
Could probably do with better parameter names than `in1` and `in2`. Marginally better perhaps is `query` and `values`?

endolith commented:
Agreed, but not sure what the best names are.  Conceptually, `in2` is a set, not an (ordered) array

eric-wieser commented:
`item`, `item_set`?

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -399,70 +400,74 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+        Default is False. ``np.isin(a, b, invert=True)`` is equivalent
+        to (but is faster than) ``np.invert(isin(a, b))``.
 
         .. versionadded:: 1.8.0
```
This versionadded is no longer relevant.

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -399,70 +400,74 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
-    python keyword `in`, for 1-D sequences. ``in1d(a, b)`` is roughly
-    equivalent to ``np.array([item in b for item in a])``.
+    `isin` can be considered as an element-wise function version of the
+    python keyword `in`. ``in1d(ar1, ar2)`` is roughly equivalent to
```
should be `isin`, not `in1d`

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -399,70 +400,74 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
-    equivalent to ``np.array([item in b for item in a])``.
+    `isin` can be considered as an element-wise function version of the
+    python keyword `in`. ``in1d(ar1, ar2)`` is roughly equivalent to
+    ``in1d = np.vectorize(lambda x: x in ar2); in1d(ar1)``.
```
I would keep the original example, which I think is a little clearer than using `np.vectorize`.

brsr replied:
The original example is not valid for multidimensional arrays.

shoyer replied:
Sure, but you can still use it to describe how this function works on 1D arrays.

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -380,16 +381,16 @@ def setxor1d(ar1, ar2, assume_unique=False):
     return aux[flag2]
 
-def in1d(ar1, ar2, assume_unique=False, invert=False):
+def isin(ar1, ar2, assume_unique=False, invert=False):
```
Definitely consider more meaningful names here.

shoyer reviewed (commented):
It occurs to me that another possible name for this function is `contains`, copying the name from `operator.contains`. That has the arguments backwards, however, which would entail flipping the arguments here, too.

Maybe `isin(query, index)` for the argument names?

brsr committed:
Fix comments on pull request

brsr commented:
Went with `elements` and `test_elements` to replace `ar1` and `ar2`. It fits with the description in the docstring: "Test whether each element of an array is present in a second array." I felt that calling `ar2` a "set" would be confusing, because the function has counterintuitive behavior if `ar2` is a standard Python set.

endolith commented:
> It occurs to me that another possible name for this function is contains, copying the name from operator.contains. That has the arguments backwards, however, which would entail flipping the arguments here, too.

To me, that would mean one array is a sub-array of the other, like `contains([[1,2,3],[4,5,6],[7,8,9]], [[1,2],[4,5]])` would be True.

https://docs.python.org/2/reference/expressions.html#membership-test-operations calls the arguments `x` and `s`, and calls `s` a "collection".  The first argument could be some synonym of `testee` or `subject`, but nothing seems obviously good to me.

eric-wieser commented:
I'd argue the name should still be `contains`, because your intuition is inconsistent with `operator.contains`: `operator.contains([1,2,3], [2,3])` is false

brsr commented:
`operator.contains` and `numpy.contains` would have inconsistent behavior. Compare:

    >>> operator.contains([1, 2, 3], 2)
    True
    >>> operator.contains([1, 2, 3], [2, 3])
    False
    >>> operator.contains([[1, 2], [3, 4], [5, 6]], [3, 6])
    False

versus

    >>> np.in1d(2, [1, 2, 3])
    array([ True], dtype=bool)
    >>> np.in1d([2, 3], [1, 2, 3])
    array([ True,  True], dtype=bool)
    >>> np.in1d([3, 6], [[1, 2], [3, 4], [5, 6]])
    array([ True,  True], dtype=bool)

eric-wieser commented:
I'm assuming that you mean `isin` not `in1d` there. 

Your second case is not really any more inconsistent than:

    >>> 1 == [1, 2]
    False
    >>> 1 == np.array([1, 2])
    array([ True,  False], dtype=bool)

I think that whatever we name it, differences in behaviour between `isin` and `in` will cause surprises, and that's not really any different to `operator.contains` vs `numpy.contains` causing those same surprises.

shoyer commented:
I guess there are enough differences here that `isin` could make more sense. I like that the name is shorter and is consistent with pandas.

seberg commented:
`__contains__` has bad behaviour in any case :/, it would be nice to deprecate the more insane stuff there.

eric-wieser commented:
@seberg: Is the behaviour documented anywhere? The first three google results simply tell me it means `in` without a link to what that means, and the third is a StackOverflow post saying _"Don't try to use __contains__ in NumPy. It does crazy, useless shit"_

seberg commented:
Reading the code of it is the easiest. Basically, it works as long as the lhs is a single element. If it is not a single element, things likely won't make much sense.

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -380,106 +381,130 @@ def setxor1d(ar1, ar2, assume_unique=False):
+    Returns a boolean array of the same length as `elements` that is True
+    where an element of `elements` is in `test_elements` and False otherwise.
+
+    See `numpy.isin` for more details.
```
I'm a little nervous about deleting all the parameters on this method, including the notes on when they were added. Can we keep them around, even though it's slightly redundant?

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -380,106 +381,130 @@ def setxor1d(ar1, ar2, assume_unique=False):
     return aux[flag2]
 
-def in1d(ar1, ar2, assume_unique=False, invert=False):
+def isin(elements, test_elements, assume_unique=False, invert=False):
```
It's not worth the hassle of changing argument names on an effectively deprecated function and potentially breaking some users. Let's only use the new arguments names for `isin`.

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -380,106 +381,130 @@ def setxor1d(ar1, ar2, assume_unique=False):
-    ar2 : array_like
-        The values against which to test each value of `ar1`.
+    test_elements : array_like
+        The values against which to test each value of `elements`.
```
Note that this argument is flattened.

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -380,106 +381,130 @@ def setxor1d(ar1, ar2, assume_unique=False):
-    in1d : (M,) ndarray, bool
-        The values `ar1[in1d]` are in `ar2`.
+    isin : ndarray, bool
+        The values `elements[isin]` are in `test_elements`.
```
Note that this has the same shape as `elements`.

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -380,106 +381,130 @@ def setxor1d(ar1, ar2, assume_unique=False):
-    # Ravel both arrays, behavior for the first array could be different
-    ar1 = np.asarray(ar1).ravel()
-    ar2 = np.asarray(ar2).ravel()
+    # Ravel both arrays, will reshape result later to match elements
```
Even though `isin` should replace most uses of `in1d`, I think this code is actually easier to understand when a separate helper function is used for the 1D functionality, in the style of `unique`/`_unique1d` in this file. So I would keep the core logic in `in1d`.

shoyer reviewed (commented):
This is very close -- would love your help to get this across the finish line!

This should also be mentioned in the docs (`doc/source/reference/routines.set.rst`) and the release notes (`doc/release/1.13.0-notes.rst`).

brsr commented:
So it sounds like we want to leave `in1d` as it is in 1.12 and just add the new function `isin` that reshapes the result of `in1d`?

shoyer commented:
@brsr You're certainly welcome to do any internal clean-up that is warranted, but from a public API perspective I would indeed keep `in1d` mostly changed. I would certainly search for references to `in1d` in the documentation and replace most of those with `isin` as the preferred solution, but we want to keep `in1d` around in an intact state to maximize backwards compatibility.

brsr committed:
keep in1d mostly the same, changes in isin now

brsr committed:
screwed up the whitespace

brsr committed:
docs, convert elements to array in case it isn't already

brsr commented:
I believe I've caught all the instances of `in1d` in the docs. The docs for `where` actually demonstrate the exact use case that I wanted this function for. Let me know what else needs tweaking.

homu commented:
:umbrella: The latest upstream changes (presumably #8446) made this pull request unmergeable. Please resolve the merge conflicts.

charris committed:
Merge branch 'master' into master

charris committed:
Merge branch 'master' into master

homu commented:
:umbrella: The latest upstream changes (presumably #8795) made this pull request unmergeable. Please resolve the merge conflicts.

brsr committed:
Merge branch 'master' into master

eric-wieser commented on doc/release/1.13.0-notes.rst:
```diff
@@ -132,6 +138,7 @@ This means that, e.g., it is now possible to do the following:
     [array([[ 1. ,  1. , -0.5], [ 1. ,  1. , -0.5]]), 
      array([[ 2. ,  2. ,  2. ], [ 2. ,  1.7,  0.5]])]
 
+
```
Looks like the merge injected an extra blank lin

brsr replied:
While I'm there, should it be using tildes or hyphens as a vertical line? The entries in the new features section aren't consistent.

eric-wieser replied:
Looking at old release notes, `~` looks correct

eric-wieser commented:
>  felt that calling ar2 a "set" would be confusing, because the function has counterintuitive behavior if ar2 is a standard Python set.

Does it? What happens when you pass a set then?

brsr committed:
removing extra line

brsr commented:
What the issue with `set`s comes down to is that if you do `np.array(set(...))`, you don't get an array of the values in that set, you get a one-element object array with that set as the sole element.

    >>> s = {1, 2, 3}
    >>> np.array(s)
    array({1, 2, 3}, dtype=object)

in1d converts its arguments to arrays using np.array, so it winds up looking for something equal to the set as an object, not the set's contents, as an element of the array.

    >>> a = np.arange(5)
    >>> np.in1d(a, s)
    array([False, False, False, False, False], dtype=bool)
    >>> b = np.empty(5, dtype=object)
    >>> b[2] = {3, 2, 1}
    >>> np.in1d(b, s)
    array([False, False,  True, False, False], dtype=bool)

Since `isin` is a brand-new part of the API, I could make it handle `set`s and set-types the expected way. On the other hand, should `array` handle set-types differently? (That would have to be a different pull request.)

(I still don't want to call the second argument `set` because that implies you have to give it a `set`, not an array-like.)

brsr committed:
support iterables that aren't array_like

Add support in isin for test_elements to be any iterable, not just an array_like. Also, clean up the docstring.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -518,21 +518,23 @@ def isin(elements, test_elements, **kwargs):
     array([0, 6])"""
     elements = np.array(elements)
+    if has_attr(test_elements, '__iter__') and not has_attr(test_elements, '__getitem__'):
+        test_elements = np.array([i for i in test_elements])
```
I think we should this path for `dict` too, so that `'a' in dict(a=1)` matches `np.isin('a', dict(a=1))`. So maybe just `if not hasattr(test_elements, '__array__')`, to just ignore things that already know how to become arrays?

eric-wieser reviewed (commented):
That last commit is a nice idea - although is this faster than something like `np.vectorize(operator.contains)` in that case? Seems like we should be able to take advantage of `set.__contains__` being O(1) here

brsr committed:
it's hasattr not has_attr

brsr committed:
check for __array__ instead

shoyer commented on numpy/lib/arraysetops.py:
```diff
@@ -481,6 +485,59 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    array([0, 6])"""
+    elements = np.array(elements)
+    if not hasattr(test_elements, '__array__'):
+        test_elements = np.array([i for i in test_elements])
```
Instead of the list comprehension, just cast to `list`, e..g, `np.array(list(test_elements))`.

brsr committed:
replace list comprehension

shoyer commented:
@eric-wieser 

> That last commit is a nice idea - although is this faster than something like np.vectorize(operator.contains) in that case? Seems like we should be able to take advantage of set.__contains__ being O(1) here

This is sometimes true, but not always, because there is significant Python overhead in using Python sets.

Ideally we would use a simple typed hashtable for these operations.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -533,7 +533,7 @@ def isin(elements, test_elements, **kwargs):
     array([0, 6])"""
     elements = np.array(elements)
-    if hasattr(test_elements, '__iter__') and not hasattr(test_elements, '__getitem__'):
+    if not hasattr(test_elements, '__array__'):
```
Probably needs a comment explaining why this is here

brsr committed:
add comment

seberg commented:
I have some questions:
1. I want to just ask whether we really want non-array like hacks in this function. On the one hand, I understand the sentiment, but I also fear it might create oddities.
2. I personally don't like the `**kwargs`. Sure, it makes keeping the functions in sync easier, but it also makes the doc less readable in my opinion. In some projects it makes sense, in numpy I am not sure.
3. I don't mind the new name, not sure if we could not actually just change the old function, but it certainly is safer/quicker, could think of deprecating the other one (in documentation), since I would argue that there is rarely to never a reason to prefer it.
4. It seems to me like it might make sense to add an `axis` argument for the second array, it would default to `None` to search the whole second array. OTOH, I somewhat think that for these type of functions we have it rarely or never implemented and we would have to discuss how exactly to handle it.

Don't feel much about anything, just trying to point out that some of these things may not be completely obvious.

brsr committed:
Responding to seberg's comments

Replace kwargs with argument names, add text in in1d deprecating
function

brsr commented:
re 1: I'm not wedded to doing this in the `isin` function. The easiest answer might be to add a note in the docstring saying "if you want to use a set, use list(set) instead". Anyone else have an opinion here?
re 4: This would be neat, but I'm not sure how to implement it. Maybe it should be a separate enhancement.

shoyer commented:
+1 to sticking with NumPy's usual way of handling non-array arguments (i.e., casting with `np.asarray`). We can document explicitly casting sets to lists.

brsr committed:
Removing special handling for sets

eric-wieser commented:
Right now:

1.  `isin(x, list({1, 2, 3})` means `x in [1, 2, 3]`
2.  `isin(x, np.array({1, 2, 3}))` means `x in `[{1, 2, 3}]`
3.  `isin(x, {1, 2, 3})` means the same as 2.

Is there an argument for making case 3 throw an exception, and require users to use one of the other explicit forms, since it is likely to cause confusion?

seberg commented:
I think we have an agreement that the last case should warn/error, but
the question is if you want to do it here, or if we want to finally
figure out how to do it in general for `np.asarray(weird_object)` ->
please provide `dtype=object` explicitly.

brsr commented:
@eric-wieser, I'm not following your reasoning. Using the example from the docs, `isin` is equivalent to `np.array([item in b for item in a])` if `a` and `b` are 1-D arrays. Expressing that in terms of the names currently used in `isin` would be something like `np.array([element in test_elements for element in elements])`. `elements` is plural because the function tests each singular `element` in `elements`.

eric-wieser commented:
> elements is plural because the function tests each singular element in elements

Right, but in numpy by default everything is plural, so that's not really helpful.

For instance, `np.add(a, b)`:

* When passed 0d arrays, does `a + b`, 
* When passed 1d arrays, does `[ai + bi for ai, bi in zip(a, b)]`
* When passed 2d arrays, does `[[aij + bij for aij, bij in zip(ai, bi)] for ai, bi in zip(a, b) ]`

It makes the most sense to document this as _"calculates `a + b`, broadcasting over a and b"_


So for `np.isin(a, bs)`:

* When passed 0d `a`, does `a in bs`, 
* When passed 1d arrays, does `[ai in bs for ai in a]`
* When passed 2d arrays, does `[[aij in bs for aij in ai] for ai in a ]`

Let's be consistent here - we should always describe the function in terms of its lowest-dimensional generalization. So it makes the most sense to document this as _"calculates `a in bs`, broadcasting over `a`"_

eric-wieser commented:
Also, we already bikeshedded this, but...

I find it very hard not to read `np.isin` as `i sin`, as python sets a precedent with `__iadd__`, and I'm used to seeing `np.sin`.

I know we've picked it for pandas compatibility, but as it's not a method of `ndarray`, this doesn't help with duck typing.

How do people feel about `np.is_in`?

shoyer commented:
> How do people feel about `np.is_in`?

I can see your point, but it looks very stilted to me, so I still prefer `isin`, which looks like a shortened version of the builtin name `isinstance`. In context, I don't think I would be confused, because the usage is so different: `np.isin` returns a boolean array typically used for indexing, rather than floats like `np.sin`.

brsr committed:
Renaming elements to element

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -516,7 +516,7 @@ def isin(elements, test_elements, assume_unique=False, invert=False):
     -------
     isin : ndarray, bool
-        Has the same shape as `elements`. The values `elements[isin]`
+        Has the same shape as `element`. The values `elements[isin]`
```
Missed an `s` here

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -527,36 +527,36 @@ def isin(elements, test_elements, assume_unique=False, invert=False):
+
+    If `test_elements` is a set (or other non-sequence collection) it will
+    be converted to an object array with one element, rather than an array
+    of the values contained in `test_elements`. Converting the set to
```
Probably worth pointing out that this is just the `np.array` constructor doing its thing, and is not something weird this function chose to do. Mentioning "`array` constructor" somewhere should do the job

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -527,36 +527,36 @@ def isin(elements, test_elements, assume_unique=False, invert=False):
-    elements = np.array(elements)
-    return in1d(elements, test_elements, assume_unique=assume_unique,
-                invert=invert).reshape(elements.shape)
+    element = np.array(element)
```
Should be `asarray`, or `asanyarray` if you think this will worked for masked arrays

eric-wieser commented on numpy/ma/extras.py:
```diff
@@ -1201,9 +1201,9 @@ def isin(elements, test_elements, assume_unique=False, invert=False):
-    elements = ma.array(elements)
-    return in1d(elements, test_elements, assume_unique=assume_unique,
-                invert=invert).reshape(elements.shape)
+    element = ma.array(element)
```
Again, `asarray` or `asanyarray`

eric-wieser commented on numpy/ma/tests/test_extras.py:
```diff
@@ -1435,6 +1435,24 @@ def test_setxor1d(self):
+        ec[0, 0, 1] = True
+        ec[0, 2, 3] = True
+        c = isin(a, b)
+        assert_array_equal(c, ec)
```
Can we have a test here for `type(c)`? Should it be `MaskedArray` or `array`.

Also, a test of what happens if you call `np.isin` instead of `np.ma.isin` might be interesting as well.

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -77,6 +77,23 @@ def test_ediff1d(self):
+        ec[1, 2, 2] = True
+        c = isin(a, b)
+        assert_array_equal(c, ec)
+
```
Needs a test for 0d as well

eric-wieser commented:
you might be able to simplify your tests with something like:

```
isin_v = np.vectorize(lambda a, b: a in b, excluded={'b'})
assert_equal(isin_v(a, b), isin(a, b))
```

brsr committed:
eric-wieser's comments

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -79,20 +79,36 @@ def test_ediff1d(self):
+        #both args
+        assert_isin_equal(5, 6)
+        #zero-d array-like as...
+        x = []
```
`[]` is 1d with shape `(0,)`, not 0d with shape `()`.

I'm thinking testing things like `np.isin(somearr, np.array(1))`, or `np.isin(np.array(1), somearr)`

eric-wieser commented on numpy/ma/tests/test_extras.py:
```diff
@@ -1435,6 +1435,28 @@ def test_setxor1d(self):
+        a = array(a, mask=mask)
+        values = [0, 10, 20, 30,  1,  3, 11, 22, 33]
+        mask   = [0,  1,  0,  1,  0,  1,  0,  1,  0]
+        b = array(values, mask=mask)
```
Or maybe even just
```
array(data = [0, 10, 20, 30,  1,  3, 11, 22, 33],
      mask = [0,  1,  0,  1,  0,  1,  0,  1,  0])
```

Better already with the alignment

brsr committed:
Fixes to tests

eric-wieser commented on numpy/ma/extras.py:
```diff
@@ -1137,15 +1137,22 @@ def setxor1d(ar1, ar2, assume_unique=False):
 
+    This function has been deprecated: use `isin` instead.
+
+    Deprecated since version 1.13.0.
```
Should use the `.. deprecated:: 1.13.0` syntax. Look around at other files to see how it's done

brsr replied:
For what it's worth, [numpy/HOWTO_DOCUMENT.rst.txt](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections) says something different about deprecation warnings.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -380,13 +381,18 @@ def setxor1d(ar1, ar2, assume_unique=False):
 
+    This function has been deprecated: use `isin` instead.
+
+    Deprecated since version 1.13.0.
```
Same here

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -77,6 +77,42 @@ def test_ediff1d(self):
+        c = [(9, 8), (7, 6)]
+        d = (9, 7)
+        assert_isin_equal(c, d)
+        #zero-d scalar as:
```
These tests are good, but tests for 0d arrays are also important, and usually catch more bugs.

Relatedly - should `isin(1, 1)` return `np.array(True)` or `np.bool_(True)`?

Also, I think the comments labelling which arg is a scalar are a bit much. This'd be more readable with just the section headings, and some space between ssections

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -411,6 +417,8 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
     --------
+    isin                  : Version of this function that preserves the
+                            shape of ar1.
     numpy.lib.arraysetops : Module with a number of other functions for
```
"See also: The module you're looking at right now" is a little weird. @brsr, you don't need to worry about it in this PR though.

brsr committed:
Docstrings, further expanding test

brsr committed:
spacing

brsr committed:
Actual zero-d array

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -481,6 +488,78 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+        can speed up the calculation.  Default is False.
+    invert : bool, optional
+        If True, the values in the returned array are inverted (that is,
+        False where an element of `ar1` is in `ar2` and True otherwise).
```
These names are out of date. I'd probably ditch this wording all together, and explain it in terms of either `not in`.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -481,6 +488,78 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    array([[ True, False],
+           [ False, True]], dtype=bool)
+    >>> element[mask]
+    array([0, 6])"""
```
nit: Extra newline in this docstring

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -481,6 +488,78 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    it will be converted to an object array with one element, rather than an
+    array of the values contained in `test_elements`. This is a consequence of
+    the `array` constructor's way of handling non-sequence collections.
+    Converting the set to a list usually gives the desired behavior.
```
This might be a good thing to show an example of, for people skipping the notes

eric-wieser commented on numpy/lib/info.py:
```diff
@@ -147,6 +147,8 @@
 in1d             Test whether elements in a 1D array are also present in
                  another array.
+isin             Test whether elements in an array are also present in
+                 another array, preserving shape of first array.
```
Maybe "test whether each element of one array is present anywhere within another"?

No strong feelings here

brsr committed:
More docstring changes

eric-wieser commented on numpy/lib/info.py:
```diff
@@ -147,8 +147,8 @@
                  another array.
-isin             Test whether elements in an array are also present in
-                 another array, preserving shape of first array.
+isin             Test whether each element of one (possibly multidimensional) 
```
Why just possibly? I think you can put `ND` before both of these arrays, to make it explicit compared to its surrounding `1D` functions

eric-wieser reviewed (approved):
Looks pretty good to me now - nice job!

brsr committed:
clean up function listing

remove 1D from section description, since they're not all 1D. match order in the list at top of arraysetops.py, where all the ND functions are first and all the 1D ones are after.

brsr committed:
Update 1.13.0-notes.rst

homu commented:
:umbrella: The latest upstream changes (presumably #8996) made this pull request unmergeable. Please resolve the merge conflicts.

brsr committed:
Merge branch 'master' into master

eric-wieser commented on doc/release/1.13.0-notes.rst:
```diff
@@ -22,6 +22,8 @@ Dropped Support
 Deprecations
 ============
 
+* ``in1d`` is now deprecated in favor of ``isin``.
```
I don't think we should mark things as deprecated without also raising a `DeprecationWarning`.

I think it would be best to not deprecate `in1d` in this PR, so that merging this otherwise nice work can happen in time for 1.13.

You can always make a new PR that adds deprecation once this is merged

shoyer replied:
This may be my fault: I may have said something like to suggest that we should deprecate in1d, but I was being sloppy with my language. I don't think we can (or should) ever raise a DeprecationWarning. `in1d` works just fine, albeit not with as nice an API as `isin`, and is used all over the place, so there is absolutely no good reason to remove it.

What we can and should do is *discourage* using `in1d` in favor of `isin` for new code. Instead of saying `in1d` is deprecated, its docstring should say something like: "We recommend using `isin` instead of `in1d` for new code".

eric-wieser commented:
Will need a rebase at some point, but if you're happy for all the commits to be squashed into one, then I can do that for you via the github UI

brsr commented:
@eric-wieser , squashing is fine with me.

brsr committed:
discouraged, not deprecated

eric-wieser commented:
Thanks @brsr!
